getwd()
source("scripts/_init.R")

distinct_samples <- 
  read_sample_sheet(inputfile, inputtab, runs, divider = "Riaz") %>% 
  distinct(name, fwd, rev)
samples <- distinct_samples %>% pull(name)
samples_1 <- paste0(samples, "_rep_1")
samples_rep <- samples_1
samples_extfrag <- paste0(samples_1, ".extendedFrags.fastq")
logfile <- "cutadapt.log"

p <- get_primers("Riaz")

cat(sep = "\n", 
    "#!/bin/bash",
    paste0("cd ~/miseq/", runname),
    paste0("mkdir ", "0_adapters_removed"),
    "\n\n",
    "#CUT ADAPTERS\n--------------\n",
    paste(collapse = "\n\n",
      paste0("echo '\n\nremoving adapters from ", samples_rep, "'\n",
             "cutadapt ",
             " -a ", p['revrc'],
             " -A ", p['fwdrc'],
             " -o ", "./0_adapters_removed/", samples_rep, "_F.fq",
             " -p ", "./0_adapters_removed/", samples_rep, "_R.fq",
             " ./demux/", samples_rep, "_F.fq", 
             " ./demux/", samples_rep, "_R.fq",
             " -m ", cutadapt_min_length, 
             " -e ", allowed_primer_error_ratio)
    ),
    paste0("echo '\n#Cutting adapters:\n'", " >> ", logfile),
    paste0("\n", "ls -l ./0_adapters_removed/*.fq", " | ", 
           "awk '{print $9, $5}'", " >> ", logfile),
    "\n\n",
    "#FAST ALIGNMENT\n--------------------\n",
    paste0("mkdir ", "./1_aligned"),
    paste(collapse = "\n\n",
          "\n\necho 'aligning ", samples_rep, "'\n",
          paste0("flash ",
                 " ./0_adapters_removed/", samples_rep, "_F.fq", 
                 " ./0_adapters_removed/", samples_rep, "_R.fq",
                 " -m ", align_min_overlap, 
                 " -M ", align_max_length, 
                 " -o ", "./1_aligned/", samples_rep)),
    paste0("echo '\n#Aligning sequences:\n'", " >> ", logfile),
    paste0("\n", "ls -l ./1_aligned/*.fastq", " | ", 
           "awk '{print $9, $5}'", " >> ", logfile),
    "\n\n",
    "#PRIMER REMOVAL\n--------------------\n",
    paste0("mkdir ",  "2_trimmed"),
    "\n #trim first primer:\n\n",
    
    paste(collapse = '\n', 
          paste0("\necho 'PRIMER_TRIM: ", samples_rep, "'", " >> ", logfile,
                 "\n",
          "cutadapt ",
                 " -g ", p["fwd"], " --discard-untrimmed", 
                 " -e ", primer_removal_error_rate, 
                 " -o ", "./2_trimmed/", samples_rep, "-trimmed-5prime.fq",
                 " ./1_aligned/", samples_rep, ".extendedFrags.fastq")),
    "\n #trim second primer:\n\n",
    paste(collapse = '\n', 
          paste0("cutadapt ",
                 " -a ", p["revrc"], " --discard-untrimmed", 
                 " -e ", primer_removal_error_rate, 
                 " -o ", "./2_trimmed/", samples_rep, "-fullytrimmed.fq",
                 " ./2_trimmed/", samples_rep, "-trimmed-5prime.fq", '\n')),
    "\n\n",
    "#QUALITY FILTERING\n--------------------\n",
    paste0("mkdir ",  "3_qfilter"),
    paste(collapse = "\n", 
          paste0("\necho ", "'filtering ", samples_rep, "'\n",
                 "vsearch ", " -fastq_stats", 
                 " ./2_trimmed/", samples_rep, "-fullytrimmed.fq",
                 " -log ", "./3_qfilter/", samples_rep, "_fastqstats.log",
                 "\n",
                 "vsearch", " -fastq_filter", 
                 " ./2_trimmed/", samples_rep, "-fullytrimmed.fq",
                 " -fastaout ", 
                 " ./3_qfilter/", samples_rep, "_merged_filtered.fa",
                 " -fastq_maxee ", quality_filtering_error, "\n")),
    "\n\n",
    "#ADD SAMPLE METADATA\n--------------------\n",
    paste0("mkdir ",  "4_obi_input"),
    paste(collapse = "\n", 
          paste0("awk '/^>/ {$0=$1 \" sample='", samples_rep, "';\"}1'",
                 " ./3_qfilter/", samples_rep, "_merged_filtered.fa",
                 " > ", "./4_obi_input/", 
                 samples_rep, "_merged_filtered_renamed.fa", "\n")),
    "\n\n",
    "#JOIN FASTAS\n--------------------\n",   
    paste(collapse = "\n",
          paste0("cat ", 
                 "./4_obi_input/", samples_rep, 
                 "_merged_filtered_renamed.fa" , 
                 " >> ", "./4_obi_input/allsamples.fa"))
)
